x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1945,14 +1945,19 @@ private void reloadSharedSecretsForJwtRealms(Settings settingsWithKeystore) {
      * See {@link TransportReloadRemoteClusterCredentialsAction} for more context.
      */
     private void reloadRemoteClusterCredentials(Settings settingsWithKeystore) {
+        // Using `settings` instead of `settingsWithKeystore` is deliberate: we are not interested in secure settings here
+        if (DiscoveryNode.isStateless(settings)) {
+            // Stateless does not support remote cluster operations. Skip.
+            return;
+        }
+
         final PlainActionFuture<ActionResponse.Empty> future = new PlainActionFuture<>();
         getClient().execute(
             ActionTypes.RELOAD_REMOTE_CLUSTER_CREDENTIALS_ACTION,
             new TransportReloadRemoteClusterCredentialsAction.Request(settingsWithKeystore),
             future
         );
-        assert future.isDone() : "expecting local-only action call to return immediately on invocation";
-        future.actionGet(0, TimeUnit.NANOSECONDS);
+        future.actionGet(10, TimeUnit.SECONDS);
```

Comment:
I'm considering a higher value here since we will be re-building more than one connection. We should specify in the docs that customers should only update one (or few) credentials at a time (if they are adding or removing creds). If they add 100 new creds in one go and some of the connections are slow it can very well take far longer than 10s.

Reply:
I'm not sure we want a timeout here. If the timeout is hit it will not cancel the underlying work and other than returning sooner not sure what that buys us.  It only encourages calling the API again which can compound the already unhealthy scenerio (and we will never be able to guess a correct upper bound unless it is huge). If anything the per-connection could have a timeout/retry ...but if does not already i don't see a need to add one now.  i.e. use the explicit listener variant instead of the implicit listener used in the future to make the call to the action.

Reply:
Good point. I like the idea of moving the timeout down to the per-connection listeners. The gnarly part is that we will have a timeout whether we want to or not for requests going through the cloud proxy. 

> use the explicit listener variant instead of the implicit listener used in the future to make the call to the action

We don't have access to the "parent" action listener of the main reload secure settings call, so without the "implicit" listener used in the future, we can't make this a blocking call. I think having it be a blocking call is useful though and mirrors what we do with a cluster setting update (if you update settings related to a remote connection, the update settings call blocks until the connection is re-built or times out). We could make this a "fire and forget", i.e., kick off connection rebuilds but don't wait for them to complete. IMO that's less user-friendly though. Ideally, I'd want to know that the connection rebuilt successfully on the reload call, instead of having to poll it out of band somehow. 

I need to think on this some more, but I'm leaning towards pushing the timeout down to the per-connection listeners. That way we can log, per connection, if the rebuild timed out and instruct users via the log line (and docs) to remedy by updating cluster settings explicitly. We also avoid one single bad connection "hogging" the request. 

FWIW, calling the reload API again on a timeout is okay. For credentials that were successfully stored we will skip re-building the connection and therefore progress.

Reply:
I ended up going with no connection timeout after an additional discussion [here](https://github.com/elastic/elasticsearch/pull/103460#discussion_r1444408370).